ccf_transaction_processor/transaction_processor/pdo_tp.cpp

```diff
@@ -137,15 +166,145 @@ namespace ccfapp
+                }
+                catch(...){
+                    return ccf::make_error(
+                        HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "Unable to parse proof data JSON");
+                }
+
+                string ias_signature = enclave_proof_data.signature;
+                string verification_report_string = enclave_proof_data.verification_report;
+
+                // verify ias report signature
+                if (!verify_ias_signature(ias_signature, attestation_policy.ias_public_key, verification_report_string)){
+                    return ccf::make_error(
+                        HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "unable to verify IAS report signature for registering enclave");
+                }
+
+                // Parse ias verification report JSON string
+                try {
+                    auto j = nlohmann::json::parse(verification_report_string);
+                    verification_report = j.get<VerificationReport>();
+                }
+                catch(...){
+                    return ccf::make_error(
+                        HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "Unable to parse IAS verification report");
+                }
+
+                string enclave_quote_body_b64 = verification_report.isvEnclaveQuoteBody;
+                vector<uint8_t> enclave_quote_body_raw = raw_from_b64(enclave_quote_body_b64);
+
+                // Verify that the verification report EPID pseudonym matches the enclave_persistent_id
+                if (in.enclave_persistent_id != verification_report.epidPseudonym) {
+                    return ccf::make_error(
+                        HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "Enclave attestation report verification Failed. Invalid epid pseudonym");
+                }
+
+                // Verify the verification report enclave quote status
+                transform(verification_report.isvEnclaveQuoteStatus.begin(), verification_report.isvEnclaveQuoteStatus.end(),
+                    verification_report.isvEnclaveQuoteStatus.begin(), ::toupper);
+                if ((verification_report.isvEnclaveQuoteStatus != OK_QUOTE_STATUS) && (verification_report.isvEnclaveQuoteStatus != GROUP_OUT_OF_DATE_QUOTE_STATUS) &&
+                     (verification_report.isvEnclaveQuoteStatus != SW_HARDENING_NEEDED_QUOTE_STATUS))  {
+                    return ccf::make_error(
+                        HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "Enclave attestation report verification Failed. Invalid quote status");
+                }
+
+                // Extract ReportData and MR_ENCLAVE from isvEnclaveQuoteBody in Verification Report
+                // The next 5 lines are copied from pservice/lib/libpdo_enclave/secret_enclave.cpp
+                sgx_quote_t* quoteBody = reinterpret_cast<sgx_quote_t*>(enclave_quote_body_raw.data());
+                sgx_report_body_t* reportBody = &quoteBody->report_body;
```

Comment:
is this a security issue? how do we know data captures a valid quoteBody? Might want to check places where you are just doing a direct cast off input.

Reply:
I am not sure I understand, enclave_quote_body is obtained from the IAS verification_report, not sure I get the "securty issue" concern.